core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/http/pipeline/stages/AsyncSigningStage.java

```diff
@@ -0,0 +1,124 @@
+import java.util.concurrent.CompletableFuture;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.core.async.AsyncRequestBody;
+import software.amazon.awssdk.core.http.ExecutionContext;
+import software.amazon.awssdk.core.interceptor.ExecutionAttributes;
+import software.amazon.awssdk.core.interceptor.SdkExecutionAttribute;
+import software.amazon.awssdk.core.internal.http.HttpClientDependencies;
+import software.amazon.awssdk.core.internal.http.InterruptMonitor;
+import software.amazon.awssdk.core.internal.http.RequestExecutionContext;
+import software.amazon.awssdk.core.internal.http.pipeline.RequestPipeline;
+import software.amazon.awssdk.core.internal.util.MetricUtils;
+import software.amazon.awssdk.core.metrics.CoreMetric;
+import software.amazon.awssdk.core.signer.AsyncRequestBodySigner;
+import software.amazon.awssdk.core.signer.AsyncSigner;
+import software.amazon.awssdk.core.signer.Signer;
+import software.amazon.awssdk.http.SdkHttpFullRequest;
+import software.amazon.awssdk.metrics.MetricCollector;
+import software.amazon.awssdk.utils.Pair;
+
+@SdkInternalApi
+public class AsyncSigningStage implements RequestPipeline<SdkHttpFullRequest,
+        CompletableFuture<SdkHttpFullRequest>> {
+
+    private final HttpClientDependencies dependencies;
+
+    public AsyncSigningStage(HttpClientDependencies dependencies) {
+        this.dependencies = dependencies;
+    }
+
+    /**
+     * Returns the response from executing one httpClientSettings request; or null for retry.
+     */
+    public CompletableFuture<SdkHttpFullRequest> execute(SdkHttpFullRequest request, RequestExecutionContext context)
+            throws Exception {
+        InterruptMonitor.checkInterrupted();
+        return signRequest(request, context);
+    }
+
+    /**
+     * Sign the request if the signer if provided and credentials are present.
+     */
+    private CompletableFuture<SdkHttpFullRequest> signRequest(SdkHttpFullRequest request,
+                                                              RequestExecutionContext context) {
+        updateInterceptorContext(request, context.executionContext());
+
+        Signer signer = context.signer();
+        MetricCollector metricCollector = context.attemptMetricCollector();
+
+        if (!shouldSign(signer)) {
+            return CompletableFuture.completedFuture(request);
+        }
+
+        adjustForClockSkew(context.executionAttributes());
+
+        if (signer instanceof AsyncSigner) {
+            AsyncSigner nonStreamingAsyncSigner = (AsyncSigner) signer;
+
+            long signingStart = System.nanoTime();
+            return nonStreamingAsyncSigner.sign(request, context.requestProvider(),
+                    context.executionAttributes()).whenComplete((r, t) ->
+                        metricCollector.reportMetric(CoreMetric.SIGNING_DURATION,
+                                Duration.ofNanos(System.nanoTime() - signingStart)));
+        }
```

Comment:
This feels brittle. It feels like it'll be easy to update one branch of the code without updating the other.

Is there a way to unify the code paths? Maybe via adapting sync signers to async signers?

Reply:
+1 that's a good idea